vllm/multimodal/utils.py

```diff
@@ -477,3 +478,36 @@ def merge_and_sort_multimodal_metadata(
         merged_hashes = None
 
     return sorted_modalities, merged_placeholders, merged_hashes
+
+
+def group_mm_inputs_by_modality(
+        mm_inputs: list["MultiModalKwargs"]) -> list[list["MultiModalKwargs"]]:
+    """Group consecutive MultiModalKwargs from mm_inputs with the same modality 
+    together into the same list for batching purpose. For MultiModalKwargs with 
+    multiple modalities, put them into their own list.
+
+    Args:
+        mm_inputs: List of MultiModalKwargs.
+
+    Returns:
+        list[list[MultiModalKwargs]]: List of list of MultiModalKwargs, each 
+        inner list contains consecutive MultiModalKwargs with same modality, or
+        one with multimodal modalities.
+    """
+    if not mm_inputs:
+        return []
+
+    def modality_group_func(mm_input: "MultiModalKwargs") -> Union[str, int]:
+        # If the input has multiple modalities, return a id as the unique key
+        # for the mm_input input.
+        if len(mm_input.modalities) > 1:
+            return id(mm_input)
+
+        # Otherwise return the modality string
+        return list(mm_input.modalities)[0]
+
+    grouped_mm_inputs = []
+    for key, group in groupby(mm_inputs, key=modality_group_func):
+        grouped_mm_inputs.append(list(group))
+
+    return grouped_mm_inputs
```

Comment:
```suggestion
    return [list(group) for _, group in groupby(mm_inputs, key=modality_group_func)]
```

Inline this into a single list comprehension

Reply:
Simplified in [d7f7990](https://github.com/vllm-project/vllm/pull/12259/commits/d7f799017c7db6a00bf6839dee033d87b03ec42c)